### PR TITLE
Basket of JIT fixes

### PR DESCRIPTION
**Pull request by ezyang**
Executive summary:

- IR pretty-printing now prints small tensors
- We now can execute Tanh, Sigmoid, AddConstant, SubConstant and Neg in JIT execution
- `verify=True` now actually does something for forwards verification (but this might get obsoleted soon)
- New `torch.random` module for the random seeding functions originally in `torch`
- New `torch.random.fork_rng` for forking the RNG state

CC @zdevito, @apaszke

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -544,6 +555,16 @@ def _flatten(obj, params=tuple()):
     return obj_vars, obj_struct
 
 
+def _clone_inputs(args):
+    def clone_input(a):
+        if isinstance(a, Variable):
+            return Variable(a.data.clone(), requires_grad=a.requires_grad, volatile=a.volatile)
+        else:
+            return a.clone()
+    return function._nested_map(lambda o: isinstance(o, Variable) or isinstance(o, Tensor),
```
Comment (fmassa) [inline torch/jit/__init__.py]: `isinstance (o, Tensor)` doesn't work for non-default tensors. It might be better to use `torch.is_tensor` instead.
But I think we should fix this in pytorch.

File: torch/random.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+def fork_rng(devices=None, enabled=True, _caller="fork_rng", _devices_kw="devices"):
+    """
+    Forks the RNG, so that when you return, the RNG is reset
+    to the state that it was previously in.
+
+    Arguments:
+        devices (iterable of CUDA IDs): CUDA devices for which to fork
+            the RNG.  CPU RNG state is always forked.  By default, fork_rng operates
+            on all devices, but will emit a warning if your machine has a lot
+            of machines, since this function will run very slowly in that case.
```
Comment (fmassa) [inline torch/random.py]: nit: I think you meant "if your machine has a lot of devices"?

File: torch/csrc/jit/ir.cpp [hunk 3]
```diff
@@ -170,8 +174,21 @@ void printAttributes(std::ostream & out, Node * n) {
         printPrimList(out,n->ss(name));
         break;
       case AttributeKind::t:
-        out << "<Tensor>";
-        break;
+        {
+          at::Tensor t = n->t(name);
+          if (t.numel() <= max_tensor_display_size) {
+            // TODO: This is awful code.  Also it doesn't work on Windows.
```
Comment (apaszke) [inline torch/csrc/jit/ir.cpp]: Why doesn't it work on windows?

File: torch/csrc/jit/ir.cpp [hunk 4]
```diff
@@ -170,8 +174,21 @@ void printAttributes(std::ostream & out, Node * n) {
...
+            std::ostringstream tensor_ss;
+            tensor_ss << t;
+            std::string tensor_s{tensor_ss.str()};
+            // Remove newlines
+            std::replace(tensor_s.begin(), tensor_s.end(), '\n', ' ');
```
Comment (apaszke) [inline torch/csrc/jit/ir.cpp]: lolol

File: torch/random.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    r"""Returns the initial seed for generating random numbers as a
+    python `long`.
+    """
+    return default_generator.initial_seed()
+
+
+_fork_rng_warned_already = False
+
+@contextlib.contextmanager
+def fork_rng(devices=None, enabled=True, _caller="fork_rng", _devices_kw="devices"):
```
Comment (apaszke) [inline torch/random.py]: what is `_devices_kw` for?

File: torch/random.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    # Internal arguments:
+    #   _caller: the function which called fork_rng, which the user used
+    #   _devices_kw: the devices keyword of _caller
+
+    if not enabled:
+        yield
+        return
+
+    if devices is None or devices is True:
+        suppress_warning = devices is True
```
Comment (apaszke) [inline torch/random.py]: what? why would devices be `True` (by identity!)? I thought it's a list of ids? If you want to test for emptiness you can't use the `is` operator

File: torch/random.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        # multiple times but a generator will be exhausted upon first traversal
+        devices = list(devices)
+
+    cpu_rng_state = torch.get_rng_state()
+    gpu_rng_states = []
+    for device in devices:
+        with torch.cuda.device(device):
+            gpu_rng_states.append(torch.cuda.get_rng_state())
+
+    yield
```
Comment (apaszke) [inline torch/random.py]: You need to
```python
try:
    yield
finally:
    # restore state
```
Otherwise you won't restore the state on exception

File: torch/jit/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/jit/__init__.py]: Thanks, this is a nice catch. This might be hard to fix in PyTorch because, semantically, a `Tensor` is whatever the default tensor is, and there is no intrinsic subclass relationship between it and any other tensor.

File: torch/csrc/jit/ir.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/jit/ir.cpp]: Windows newlines are \r\n, but this will only search replace \n. But going from one character to multi-character search replace requires exercising different C++ stuff

File: torch/jit/__init__.py (hunk 1 above)
Comment (apaszke) [reply torch/jit/__init__.py]: False. With `__instancecheck__` and a few other Python tricks you can make `torch.Tensor` do all kinds of arbitrary things 😊

Comment (fmassa) [reply torch/jit/__init__.py]: Another option would be to go into the ABC thing, but I think this is not going to happen anymore https://github.com/pytorch/pytorch/issues/399?

File: torch/random.py (hunk 5 above)
Comment (ezyang) [reply torch/random.py]: It's a private method for telling users what keyword argument they feed devices by. I didn't want to hardcode 'devices' because it might be a different kw from the user-visible interface.

File: torch/random.py (hunk 6 above)
Comment (ezyang) [reply torch/random.py]: OK, here's the back story: the code here always emits a warning when you have >1 GPU on your machine. So, what if you really did want all of the GPUs to be enabled? You might want to suppress the warning, in this case. `devices=None` has the correct semantics, but it emits warnings. I didn't want to force users to list them all out in the `devices` argument just to suppress the warning. So I made a new mode of calling (documented in the warning message), `devices=True`, which has the same semantics as `devices=None` but suppresses the warning.

We can just delete this logic and make the warning unsuppressable without specifying all devices manually if you like.

Comment (apaszke) [reply torch/random.py]: This is quite awkward. Once someone learns about the pitfalls of this function, they will always have to write `devices=True`, which is not an acceptable for any other function in our framework that takes a `devices` argument

File: torch/jit/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/jit/__init__.py]: I prefer the ABC solution.

File: torch/random.py (hunk 6 above)
Comment (ezyang) [reply torch/random.py]: OK the `devices=True` path has been axed.
